File: src/main/java/kr/devport/api/domain/common/cache/CacheKeyFactory.java
package kr.devport.api.domain.common.cache;

import kr.devport.api.domain.llm.enums.BenchmarkType;
import org.springframework.aot.hint.MemberCategory;
import org.springframework.aot.hint.RuntimeHints;
import org.springframework.aot.hint.RuntimeHintsRegistrar;
import org.springframework.context.annotation.ImportRuntimeHints;
import org.springframework.stereotype.Component;

import java.math.BigDecimal;
import java.util.Arrays;
import java.util.stream.Collectors;

/**
 * Canonical cache key factory for scoped cached reads.
 *
 * Registered as a Spring bean so SpEL expressions in @Cacheable annotations
 * can reference it via {@code @cacheKeyFactory.method()} — avoiding T() type
 * lookups that fail in GraalVM native images.
 */
@Component
@ImportRuntimeHints(CacheKeyFactory.CacheKeyFactoryRuntimeHints.class)
public class CacheKeyFactory {

    public static class CacheKeyFactoryRuntimeHints implements RuntimeHintsRegistrar {
        @Override
        public void registerHints(RuntimeHints hints, ClassLoader classLoader) {
            hints.reflection().registerType(CacheKeyFactory.class, MemberCategory.INVOKE_PUBLIC_METHODS);
        }
    }

    // ========== Article Domain Keys ==========

    public String articleListKey(kr.devport.api.domain.article.enums.Category category, int page, int size) {
        String categoryKey = category != null ? category.name() : "all";
        return String.format("%s_%d_%d", categoryKey, page, size);
    }

    public String trendingTickerKey(int limit) {
        return String.valueOf(limit);
    }

    // ========== Git Repository Domain Keys ==========

    public String gitRepoListKey(kr.devport.api.domain.gitrepo.enums.Category category, int page, int size) {
        String categoryKey = category != null ? category.name() : "all";
        return String.format("%s_%d_%d", categoryKey, page, size);
    }

    public String trendingGitReposKey(int page, int size) {
        return String.format("%d_%d", page, size);
    }

    public String gitReposByLanguageKey(String language, int limit) {
        String normalizedLanguage = normalizeString(language);
        return String.format("%s_%d", normalizedLanguage, limit);
    }

    // ========== LLM Domain Keys ==========

    public String llmLeaderboardKey(
        BenchmarkType benchmarkType,
        String provider,
        String creatorSlug,
        String license,
        BigDecimal maxPrice,
        Long minContextWindow
    ) {
        String benchmarkKey = benchmarkType.name();
        String providerKey = normalizeString(provider);
        String creatorKey = normalizeString(creatorSlug);
        String licenseKey = normalizeString(license);
        String priceKey = maxPrice != null ? maxPrice.toPlainString() : "null";
        String contextKey = minContextWindow != null ? minContextWindow.toString() : "null";

        return String.format("%s_%s_%s_%s_%s_%s",
            benchmarkKey, providerKey, creatorKey, licenseKey, priceKey, contextKey);
    }

    public String allBenchmarksKey() {
        return "all";
    }

    // ========== Normalization Utilities ==========

    private String normalizeString(String value) {
        if (value == null || value.trim().isEmpty()) {
            return "all";
        }
        return value.trim().toLowerCase();
    }

    private String normalizeList(String... values) {
        if (values == null || values.length == 0) {
            return "all";
        }

        String normalized = Arrays.stream(values)
            .filter(v -> v != null && !v.trim().isEmpty())
            .map(String::trim)
            .map(String::toLowerCase)
            .sorted()
            .collect(Collectors.joining(","));

        return normalized.isEmpty() ? "all" : normalized;
    }
}

File: support/cache/src/main/kotlin/kr/devport/api/domain/common/cache/CacheKeyFactory.kt
package kr.devport.api.domain.common.cache

import org.springframework.stereotype.Component
import java.math.BigDecimal

/**
 * Canonical cache key factory for scoped cached reads. A Spring bean so SpEL in @Cacheable can
 * reference it via `@cacheKeyFactory.method(...)`. Domain enums are passed in as their String name
 * (SpEL `#category?.name()`) so this shared factory has no per-domain dependency.
 */
@Component
class CacheKeyFactory {
    // ===== Article =====

    fun articleListKey(
        category: String?,
        page: Int,
        size: Int,
    ): String {
        val categoryKey = if (!category.isNullOrBlank()) category else "all"
        return "%s_%d_%d".format(categoryKey, page, size)
    }

    fun trendingTickerKey(limit: Int): String = limit.toString()

    // ===== Git repository =====

    fun gitRepoListKey(
        category: String?,
        page: Int,
        size: Int,
    ): String {
        val categoryKey = if (!category.isNullOrBlank()) category else "all"
        return "%s_%d_%d".format(categoryKey, page, size)
    }

    fun trendingGitReposKey(
        page: Int,
        size: Int,
    ): String = "%d_%d".format(page, size)

    fun gitReposByLanguageKey(
        language: String?,
        limit: Int,
    ): String = "%s_%d".format(normalizeString(language), limit)

    // ===== LLM =====

    fun llmLeaderboardKey(
        benchmarkType: String?,
        provider: String?,
        creatorSlug: String?,
        license: String?,
        maxPrice: BigDecimal?,
        minContextWindow: Long?,
    ): String {
        val benchmarkKey = benchmarkType ?: "null"
        val providerKey = normalizeString(provider)
        val creatorKey = normalizeString(creatorSlug)
        val licenseKey = normalizeString(license)
        val priceKey = maxPrice?.toPlainString() ?: "null"
        val contextKey = minContextWindow?.toString() ?: "null"
        return "%s_%s_%s_%s_%s_%s".format(benchmarkKey, providerKey, creatorKey, licenseKey, priceKey, contextKey)
    }

    fun allBenchmarksKey(): String = "all"

    private fun normalizeString(value: String?): String {
        if (value.isNullOrBlank()) {
            return "all"
        }
        return value.trim().lowercase()
    }
}
